### PR TITLE
GH-35658: [Packaging] Sync conda recipes with feedstocks

### LINKED ISSUE #35658: [Packaging] Sync with conda-forge feedstocks
### Describe the bug, including details regarding any error messages, version, and platform.

Regular updates to catch updates after 12.0 and new pinnings that arrived in the meantime

### Component(s)

Packaging

### PR DESCRIPTION
**Pull request by h-vetinari**
Corresponds to https://github.com/conda-forge/arrow-cpp-feedstock/pull/1053 for pyarrow (one failing test on our infra that needs debugging), as well as the state of the feedstock for r-arrow after https://github.com/conda-forge/r-arrow-feedstock/pull/65
* Closes: #35658

### CODE REVIEW

**Comment by h-vetinari** [comment]
@github-actions crossbow submit -g conda

**Comment by h-vetinari** [comment]
Interestingly, the failure of `test_total_bytes_allocated` on linux from https://github.com/conda-forge/arrow-cpp-feedstock/pull/1053 is not present here. But a weird error on windows:
```
E   pyarrow.lib.ArrowInvalid: Cannot locate timezone 'UTC': Timezone database not found at "C:\Users\VssAdministrator\Downloads\tzdata"
```

**Comment by kou** [comment]
I don't know why we didn't need timezone database for Windows jobs before but we have a shell script to download timezone database:  https://github.com/apache/arrow/blob/main/ci/scripts/download_tz_database.sh

We're using the script on Windows too.

**Comment by h-vetinari** [comment]
> I don't know why we didn't need timezone database for Windows jobs before but we have a shell script to download timezone database:

Sounds like we should just add a dependency on the [`tzdata`](https://github.com/conda-forge/tzdata-feedstock) conda-forge package?

**Comment by assignUser** [comment]
> Sounds like we should just add a dependency on the [`tzdata`](https://github.com/conda-forge/tzdata-feedstock) conda-forge package?

That sound's good. Thanks for taking care of the sync work, much appreciated!

**Comment by kou** [comment]
It may not work because it seems that `tzdata` doesn't include `WindowsZones.xml`.

**Comment by h-vetinari** [comment]
> It may not work because it seems that `tzdata` doesn't include `WindowsZones.xml`.

Why does arrow depend on this in the first place? It's the first package I see that doesn't work with `tzdata`. Also, based on the error message, it seems to be looking in the wrong place? I.e. `C:\Users\VssAdministrator\Downloads\tzdata` instead of `%LIBRARY_PREFIX%` (aside from the fact that inndeed, our tzdata does not [contain](https://github.com/regro/libcfgraph/blob/master/artifacts/tzdata/conda-forge/noarch/tzdata-2023c-h71feb2d_0.json#L231-L835) `WindowsZones.xml`).

The tzdata stuff is an arrow 13 problem though. More urgently, in https://github.com/conda-forge/arrow-cpp-feedstock/pull/1053, I currently have a failing test `test_total_bytes_allocated` on arrow 12 (though not arrow 11 or before), but some commit between arrow 12 and current `main` seems to have fixed this - any ideas?

**Comment by raulcd** [comment]
As a note, our nightly builds for wheels on Windows are also failing at the moment with the same error:
```
 E   pyarrow.lib.ArrowInvalid: Cannot locate timezone 'UTC': Timezone database not found at "C:\Users\ContainerAdministrator\Downloads\tzdata"
```
* [wheel-windows-cp310-amd64](https://github.com/ursacomputing/crossbow/actions/runs/4989689010/jobs/8933937653)
* [wheel-windows-cp311-amd64](https://github.com/ursacomputing/crossbow/actions/runs/4989688097/jobs/8933935592)
* [wheel-windows-cp37-amd64](https://github.com/ursacomputing/crossbow/actions/runs/4989683949/jobs/8933925508)
* [wheel-windows-cp38-amd64](https://github.com/ursacomputing/crossbow/actions/runs/4989686872/jobs/8933932531)
* [wheel-windows-cp39-amd64](https://github.com/ursacomputing/crossbow/actions/runs/4989690690/jobs/8933941809)
The first failure is from 10 days ago since this commit was introduced:
https://github.com/apache/arrow/commit/053b5eecb582a93bc92e6793768fc66d2c5d5046

**Comment by kou** [comment]
Ah, sorry. I misread this problem.

It seems that `WindowsZones.xml` isn't related to this problem.

But tzdata path was hard-coded on Windows: https://github.com/apache/arrow/blob/main/cpp/src/arrow/vendored/datetime/tz.cpp#L265

So our code will not find tzdata installed by conda.

Could you open a new issue for tzdata related problem? We can work on this as a separated issue.

**Comment by raulcd** [comment]
The datetime vendored library seems to have had a change lately: https://github.com/apache/arrow/pull/35612/files
And is quite important for the R release to be packed for 12.0.1. I am wondering whether this issue will have to be fixed for the 12.0.1 release too.

**Comment by h-vetinari** [comment]
> 


> The tzdata stuff is an arrow 13 problem though. More urgently, in [conda-forge/arrow-cpp-feedstock#1053](https://github.com/conda-forge/arrow-cpp-feedstock/pull/1053), I currently have a failing test `test_total_bytes_allocated` on arrow 12 (though not arrow 11 or before), but some commit between arrow 12 and current `main` seems to have fixed this - any ideas?

As I mentioned, the tzdata thing is much less urgent for us in conda-forge than the issue of that failing tests on the feedstock. I now opened a separate issue so this gets (hopefully...) more traction: #35728

**Comment by jorisvandenbossche** [comment]
> Why does arrow depend on this in the first place? It's the first package I see that doesn't work with `tzdata`.

Arrow unfortunately can't work with the tz database from `tzdata`, because the underlying date library we use doesn't support the binary format, see https://github.com/apache/arrow/issues/31472 and linked issues.

In other places we skip tests that requite tz database access on Windows, so that should be done for the new scalar test as well, I assume.

**Comment by jorisvandenbossche** [comment]
I opened https://github.com/apache/arrow/pull/35735 to skip the test requiring a tz database on windows.

**Comment by h-vetinari** [comment]
> Arrow unfortunately can't work with the tz database from `tzdata`, because the underlying date library we use doesn't support the binary format, see #31472 and linked issues.

I don't mind shipping arrow's vendored `WindowsZones.xml` in a pyarrow-specific location (i.e. site-packages). I think it would be better to do that than just fail? timezone-operations are pretty common (unless I misunderstand something), so not supporting that for the conda-forge builds would suck IMO

**Comment by jorisvandenbossche** [comment]
>  so not supporting that for the conda-forge builds would suck IMO

Yes, but to be clear, we currently don't support it for any form of distribution (also not wheels), the user is required to download it manually. So while I also certainly would like to see this solved, it's not something where conda-forge is worse.

Although it might be easier to solve for conda-forge:

> I don't mind shipping arrow's vendored `WindowsZones.xml` in a pyarrow-specific location (i.e. site-packages/pyarrow; we can patch pyarrow to point to the right location).

That should be possible yes (I think it doesn't even need to be in a pyarrow-specific location, it could also be in a "proper" location for data). Currently this would require a patch in the C++ code, with https://github.com/apache/arrow/issues/35600 it should be an easier patch in the pyarrow's `__init__.py`

One issue with vendoring is that this doesn't make it easy to update the version / ensure you always have a recent tz database, also when installing an somewhat older pyarrow version.

**Comment by h-vetinari** [comment]
> One issue with vendoring is that this doesn't make it easy to update the version / ensure you always have a recent tz database, also when installing an somewhat older pyarrow version.

Absolutely, that's why it would be nice if this could be handled through a "proper" dependency like `tzdata`. I haven't dived deep into https://github.com/apache/arrow/issues/31472, but I don't understand why the format that `tzdata` has couldn't be transformed or otherwise made to work for arrow?

**Comment by jorisvandenbossche** [comment]
That's probably technically possible to implement such transformation, but someone still needs to do the work for that. A potential easier path forward is to update date.h to work with the compiled database on Windows (https://github.com/HowardHinnant/date/issues/564)

**Comment by h-vetinari** [comment]
@github-actions crossbow submit -g conda

**Comment by h-vetinari** [comment]
@github-actions crossbow submit -g conda

**Comment by h-vetinari** [comment]
@kou @pitrou @raulcd @assignUser @jorisvandenbossche 
This is green now and matches the state of the feedstocks. There are some more minor improvements on the way (a few but not all) of the test skips can be removed, but I can also put those in the next sync PR after arrow 13 is officially out.

**Review by assignUser — APPROVED** [review]
Great work, thanks @h-vetinari ! I will merge in a day or so if there are not objections

File: dev/tasks/conda-recipes/arrow-cpp/build-arrow.sh [hunk 1]
```diff
@@ -30,27 +30,21 @@ fi
...
+    # point to a usable protoc/grpc_cpp_plugin if we're cross-compiling
+    EXTRA_CMAKE_ARGS="${EXTRA_CMAKE_ARGS} -DProtobuf_PROTOC_EXECUTABLE=$BUILD_PREFIX/bin/protoc"
+    if [[ ! -f ${BUILD_PREFIX}/bin/${CONDA_TOOLCHAIN_HOST}-clang ]]; then
+        ln -sf ${BUILD_PREFIX}/bin/clang ${BUILD_PREFIX}/bin/${CONDA_TOOLCHAIN_HOST}-clang
+    fi
+    EXTRA_CMAKE_ARGS="${EXTRA_CMAKE_ARGS} -DCLANG_EXECUTABLE=${BUILD_PREFIX}/bin/${CONDA_TOOLCHAIN_HOST}-clang"
+    EXTRA_CMAKE_ARGS="${EXTRA_CMAKE_ARGS} -DLLVM_LINK_EXECUTABLE=${BUILD_PREFIX}/bin/llvm-link"
     sed -ie "s;protoc-gen-grpc.*$;protoc-gen-grpc=${BUILD_PREFIX}/bin/grpc_cpp_plugin\";g" ../src/arrow/flight/CMakeLists.txt
-    sed -ie 's;"--with-jemalloc-prefix\=je_arrow_";"--with-jemalloc-prefix\=je_arrow_" "--with-lg-page\=14";g' ../cmake_modules/ThirdpartyToolchain.cmake
+    sed -ie 's;"--with-jemalloc-prefix\=je_arrow_";"--with-jemalloc-prefix\=je_arrow_" "--with-lg-page\=16";g' ../cmake_modules/ThirdpartyToolchain.cmake
```
Comment (kou) [inline dev/tasks/conda-recipes/arrow-cpp/build-arrow.sh:47]: We can use `-DARROW_JEMALLOC_LG_PAGE=16` instead of rewriting `ThirdpartyToolchain.cmake`.

**Review by kou — APPROVED** [review]
+1

**Review by raulcd — APPROVED** [review]

File: dev/tasks/conda-recipes/arrow-cpp/build-arrow.sh (hunk 1 above)
Comment (h-vetinari) [reply dev/tasks/conda-recipes/arrow-cpp/build-arrow.sh:47]: That sounds great, I'll change that on the feedstock and in the next sync

**Pull request merged**